### ISSUE TITLE
Discrepancies in the way Taquito formats complex pairs inside and outside lists

### ISSUE THREAD

**Issue by mattou78400** [ecadlabs/taquito#1054]
Hi,

I deployed the following smart contract:
https://better-call.dev/granadanet/KT1Jgt71qfrqNjmYkcsTz4PgMMKJEgX1wCa4/operations

re deploy of:
https://better-call.dev/mainnet/KT1U1JZaXoG4u1EPnhHL4R4otzkWc1L34q3c/interact?entrypoint=addMembers

I'm trying to call the function addMembers, but I don't seem to catch how to format the input

Here's what I've tried

`
var new_member = {country : country, expires : expires, rating : rating, region : region, restricted : restricted}

      const data = [address_tz, MichelsonMap.fromLiteral(new_member)]
      // address_tz is my own wallet and is set correctly
      console.log(new_member, data, address_tz)

      const op = await Tezos.contract.at('KT1Jgt71qfrqNjmYkcsTz4PgMMKJEgX1wCa4')

      const loadedNewMember = await op.methods.addMembers(data).send();

      await loadedNewMember.confirmation()
`

Though I do seem to hit the endpoint, it tells me

![image](https://user-images.githubusercontent.com/39678117/134073534-9cfa054b-ac00-4ce8-a6c9-51758516a541.png)

What am I missing? How should I format `new_member` or `data` to hit the endpoint correctly?

It was raised in the Telegram channel that it might be a bug... But I can't figure it out
Thanks a lot!

endpoint:
![image](https://user-images.githubusercontent.com/39678117/134073807-5ee89716-1908-41ec-b802-a30f5b28acf7.png)

**Comment by claudebarde** [comment]
The parameter you pass to the contract entrypoint in your example is wrong, a Michelson `Pair` doesn't translate to a `MichelsonMap` with Taquito. In this case, because the parameter is a list, the values should be in an object, as we talked about in Telegram:
`[
   {
      0: address,
      country: country,
      expires: data,
      rating: rating,
      region: region,
      restricted: true|false
   }
]`

**Comment by mattou78400** [comment]
Thanks Claude!

I tried this, but I'm now getting this upon calling the method:
![image](https://user-images.githubusercontent.com/39678117/134472769-f085a72d-3321-483a-a7de-09c8494008fb.png)

`Error: Unable to encode storage object. TypeError: Cannot read properties of undefined (reading 'restricted')`

Any idea why "restricted" isn't recognized?

**Comment by claudebarde** [comment]
Do you have a Github repo available so we can reproduce the error?

**Comment by mattou78400** [comment]
Not really 
But to reproduce:

Fork and deploy this smart contract: 
https://better-call.dev/granadanet/KT1WVMzpceKDXBuYGZisypTmH5iK3PZoDS25/interact?entrypoint=addMembers


Try to call the function as such: 
`
var new_member = {0 : deploy_wallet_address, country : country, expires : expires, rating : rating, region : region, restricted : false}
const data = [new_member]
const op = await Tezos.contract.at('SMART_CONTRACT_ADDRESS')
const loadedNewMember = await op.methods.addMembers(data).send();
await loadedNewMember.confirmation()
`

**Comment by mattou78400** [comment]
Could you reproduce it?

**Comment by claudebarde** [comment]
Yes, I've been able to reproduce it, I am investigating the problem now :)

**Comment by claudebarde** [comment]
The error comes from the fact that the expected parameter is not the right one. Taquito generally spreads the fields of complex pairs and the expected parameter would thus be:
```
{
      0: implicit_account_address, 
      country: 1, 
      expires: new Date().toISOString(), 
      rating: 20, 
      region: 35, 
      restricted: false
}
```
However, this is not the case here and the correct parameter is:
```
{
      0: implicit_account_address, 
      1: {
            country: 1, 
            expires: new Date().toISOString(), 
            rating: 20, 
            region: 35, 
            restricted: false
        }
}
```
This operation has been created and went through with the parameter above: [Operation](https://better-call.dev/granadanet/opg/ooFpn4XdrvcLfqGNiHrsgBqhFywdyeyKQaqcQzs1WWKoSYmDskb/contents)

There are discrepancies in the way Taquito formats complex pairs inside and outside lists, this should be investigated further.

**Comment by mattou78400** [comment]
Amazing, thanks a lot, it worked for me too :)

**Comment by claudebarde** [comment]
Here is a script that showcases the different formatting Taquito expects for complex pairs inside and outside a list: https://replit.com/@claudebarde/TaquitoComplexPairInListTest#index.ts
The underlying smart contract has been recreated with the exact same parameter signature as the contract mentioned in this issue.
Such a discrepancy shouldn't exist, this should be investigated.
